src/main/java/week6/ParkingLot.java

```diff
@@ -0,0 +1,152 @@
+package week6;
+
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.logging.Logger;
+
+public class ParkingLot {
+
+  private static final int EMPTY = 0;
+  private static final Logger logger = Logger.getLogger(ParkingLot.class.getName());
+
+  /*
+   * First, we create a map of wrongly positioned cars, car no. -> position
+   * Iteratively, replace 0 with the correct car. We remove cars that are in correct pos from map
+   *
+   * If 0 becomes in the correct position but map is not empty, move 0 to first incorrect car pos
+   * and start iterating again until map is empty.
+   *
+   * Amortized Time Complexity: O(n)
+   * Space Complexity: O(n)
+   *
+   * where n is the size of the array source
+   *
+   * Assumptions:
+   *    1) len(source) = len(target)
+   *    2) arrays contain O...n, this means no duplications
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @return : List of Moves to be performed to transform source -> target
+   */
+  public static List<Move> rearrangeCars(final int[] source, final int[] target) {
+
+    List<Move> moves = new ArrayList<Move>() {
+      @Override
+      public String toString() {
+        return super.toString().replaceAll("\\[|]|(, )", "");
+      }
+    };
+
+    if (source == null || target == null || source.length != target.length) {
+      return moves;
+    }
+
+    //creates map (wrongly positioned) car id -> position and get index of the empty Box
+    HashMap<Integer, Integer> misplacedCars = new HashMap<>();
+    int emptyBox = populateMap(source, target, misplacedCars);
+
+    while (!misplacedCars.isEmpty()) {
+      //if emptyBox is in right pos, we swap the emptyBox with one of the wrong cars
+      if (target[emptyBox] == EMPTY) {
+        emptyBox = correctEmpty(misplacedCars, target, emptyBox, moves);
+      }
+      emptyBox = moveEmpty(misplacedCars, target, emptyBox, moves);
+    }
+
+    logger.info(moves.toString());
+    return moves;
+  }
+
+
+  /*
+   * Updates the map misPlacedCars : cars that needs to be moved -> current position.   *
```

Comment:
Nit: Inconsistent capitalisation of misPlacedCars (this can come back to bite you if you rename the variable later, and it doesn't catch the typo'd variable), and a lost "*" at the end of the line.